File: src/types/proposal.ts
import type { DocumentPayload, DocumentVersion } from "./workspace";
import type { DocumentSnapshotRecord } from "./snapshot";

export type { DocumentSnapshotRecord } from "./snapshot";

export interface AdapterDefinition {
  id: string;
  name: string;
  command: string;
  args: string[];
  workingDirectoryMode: string;
  environmentAllowlist: string[];
  inputMode: string;
  outputMode: string;
  capabilities: string[];
  timeoutSeconds: number;
}

export interface ProposalRecord {
  schemaVersion: number;
  id: string;
  documentId: string;
  threadIds: string[];
  adapterId: string;
  createdAt: string;
  updatedAt: string;
  status: "pending" | "accepted" | "rejected" | "stale" | "failed";
  baseContentHash: string;
  responseMode: string;
  summary: string;
  assistantMessage: string;
  updatedDocumentText: string | null;
  unifiedDiff: string | null;
  computedDiff: string;
  warnings: string[];
  resolveThreadIds: string[];
  stderr: string | null;
  errorMessage: string | null;
}

export interface TextRange {
  startByte: number;
  endByte: number;
  startLine: number;
  endLine: number;
}

export type HunkKind = "insert" | "delete" | "replace";

export interface ChangeSource {
  kind: string;
  id?: string | null;
}

export interface ReviewHunk {
  id: string;
  index: number;
  kind: HunkKind;
  oldRange: TextRange;
  newRange: TextRange;
  beforeText: string;
  afterText: string;
  headingPath: string[];
  summary?: string | null;
}

export interface ReviewChangeSet {
  id: string;
  documentId: string;
  source: ChangeSource;
  beforeHash: string;
  afterHash: string;
  baseHash?: string | null;
  hunks: ReviewHunk[];
  warnings: string[];
}

export type ProposalChangeSetResult =
  | {
      status: "ready";
      proposal: ProposalRecord;
      changeSet: ReviewChangeSet;
      documentVersion: DocumentVersion;
    }
  | {
      status: "stale";
      proposal: ProposalRecord;
      message: string;
    }
  | {
      status: "unsupported";
      proposal: ProposalRecord;
      message: string;
    };

export interface ProposalMutationResult {
  proposal: ProposalRecord;
  document: DocumentPayload | null;
  snapshot: DocumentSnapshotRecord | null;
  message: string | null;
}

export type ProposalHunkAcceptResult =
  | {
      status: "applied";
      result: ProposalMutationResult;
      appliedHunkIds: string[];
    }
  | {
      status: "conflict";
      expectedVersion: DocumentVersion;
      actualVersion: DocumentVersion;
      message: string;
    }
  | {
      status: "stale";
      proposal: ProposalRecord;
      message: string;
    };

File: src-tauri/src/models/proposal.rs
use serde::{Deserialize, Serialize};

use crate::models::document::{DocumentPayload, DocumentVersion};
use crate::models::snapshot::DocumentSnapshotRecord;

pub use margent_core::change_set::ReviewChangeSet;
pub use margent_core::proposal::ProposalRecord;

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProposalMutationResult {
    pub proposal: ProposalRecord,
    pub document: Option<DocumentPayload>,
    pub snapshot: Option<DocumentSnapshotRecord>,
    pub message: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(
    tag = "status",
    rename_all = "camelCase",
    rename_all_fields = "camelCase"
)]
pub enum ProposalChangeSetResult {
    Ready {
        proposal: Box<ProposalRecord>,
        change_set: Box<ReviewChangeSet>,
        document_version: DocumentVersion,
    },
    Stale {
        proposal: Box<ProposalRecord>,
        message: String,
    },
    Unsupported {
        proposal: Box<ProposalRecord>,
        message: String,
    },
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(
    tag = "status",
    rename_all = "camelCase",
    rename_all_fields = "camelCase"
)]
pub enum ProposalHunkAcceptResult {
    Applied {
        result: Box<ProposalMutationResult>,
        applied_hunk_ids: Vec<String>,
    },
    Conflict {
        expected_version: DocumentVersion,
        actual_version: DocumentVersion,
        message: String,
    },
    Stale {
        proposal: Box<ProposalRecord>,
        message: String,
    },
}
